### PR TITLE
[Security] updated BBUserToken API to provide ::isExpired method

### PR DESCRIPTION
**Pull request by eric-chau**
Updated ``BackBee\Security\Token\BBUserToken`` to add ``::setLifetime`` and ``::isExpired`` methods. By this way we can ask the token itself if it is expired or not so ``BBApplication::getBBUserToken`` can invalidate session if the session is expired.

ping @crouillon, what do you think about it?

(This PR aims to resolve a part of [backbee/bb-core-js issue #522](https://github.com/backbee/BbCoreJs/issues/522))

### CODE REVIEW

File: Security/Authentication/Provider/BBAuthenticationProvider.php [hunk 1]
```diff
@@ -147,7 +147,12 @@ public function authenticate(TokenInterface $token)
         }
 
         $validToken = new BBUserToken($user->getRoles());
-        $validToken->setUser($user)->setNonce($token->getNonce());
+        $validToken
+            ->setUser($user)
+            ->setNonce($token->getNonce())
+            ->setCreated(new \DateTime())
+            ->setLifetime($this->lifetime)
```
Comment (mickaelandrieu) [inline Security/Authentication/Provider/BBAuthenticationProvider.php]: \o/ yes, but I want to configure this lifetime through the configuration

```yaml
// config.yml
session:
    lifetime: 86400 // in seconds
```
What do you think ?

File: Security/Token/BBUserToken.php [hunk 2]
```diff
@@ -40,21 +40,28 @@ class BBUserToken extends AbstractToken
...
      */
-    private $_nonce;
+    private $nonce;
+
+    /**
+     * Token max lifetime (in second).
+     *
+     * @var integer
+     */
+    private $lifetime = 300;
```
Comment (mickaelandrieu) [inline Security/Token/BBUserToken.php]: 300 seconds ? the common lifetime is about 20 min I think

Comment (eric-chau) [inline Security/Token/BBUserToken.php]: Ok I will update it

File: Security/Authentication/Provider/BBAuthenticationProvider.php (hunk 1 above)
Comment (eric-chau) [inline Security/Authentication/Provider/BBAuthenticationProvider.php]: The current behavior is that it take the lifetime provided by the firewall that authenticated the token. In the ``rest_api_area`` case:

```yaml
# backbee/backbee/Config/security.yml
rest_api_area:
        pattern: ^/rest
        requirements:
            HTTP-X-API-SIGNATURE: \w+
            HTTP-X-API-KEY: \w+
        provider: public_key
        public_key: ~
        restful:
            provider: bb_user
            nonce_dir: security/nonces
            lifetime: 1800 # this one is used, 30 min by default
            use_registry: true
        anonymous: ~
```

File: Security/Token/BBUserToken.php [hunk 3]
```diff
@@ -36,25 +36,37 @@
...
      */
-    private $_nonce;
+    private $nonce;
+
+    /**
+     * Token max lifetime (in second).
+     *
+     * @var integer
+     */
+    private $lifetime = self::DEFAULT_LIFETIME;
```
Comment (eric-chau) [inline Security/Token/BBUserToken.php]: ping @mickaelandrieu, does it okay for you now?

File: Security/Authentication/Provider/BBAuthenticationProvider.php (hunk 1 above)
Comment (mickaelandrieu) [inline Security/Authentication/Provider/BBAuthenticationProvider.php]: This make sense :)

**Comment by mickaelandrieu** [comment]
Nice job @eric-chau

File: Security/Token/BBUserToken.php (hunk 3 above)
Comment (mickaelandrieu) [inline Security/Token/BBUserToken.php]: yes, thanks :)

**Comment by mickaelandrieu** [comment]
@eric-chau do you want I launch tests of this branch from my side ? Travis sounds a little bit KO today.

**Comment by eric-chau** [comment]
@mickaelandrieu I know tests are ok I tested it on local but I would like to get @crouillon feedbacks too before merging it.

File: BBApplication.php [hunk 4]
```diff
@@ -385,26 +386,19 @@ public function getContext()
...
-                if (null !== $token = $this->getContainer()->get('bb_session')->get('_security_bb_area')) {
-                    $token = unserialize($token);
 
-                    if (!($token instanceof BBUserToken)) {
-                        $token = null;
-                    }
-                }
-            }
+        if ($token instanceof BBUserToken && $token->isExpired()) {
+            $this->getRequest()->getSession()->invalidate();
```
Comment (crouillon) [inline BBApplication.php]: Could dispatch frontontroller.request.logout event ?

**Comment by eric-chau** [comment]
Review done with @crouillon this morning. Thanks @mickaelandrieu  and @crouillon  for reviewing this PR.

**Pull request merged**